### PR TITLE
fix(rbac): fix SA name in CRB subjects

### PR DESCRIPTION
**Pull request by stavros-k**
**Description**

⚒️ Fixes  https://github.com/truecharts/charts/issues/3871

**⚙️ Type of change**

- [ ] ⚙️ Feature/App addition
- [x] 🪛 Bugfix
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**

**📃 Notes:**

**✔️ Checklist:**

- [x] ⚖️ My code follows the style guidelines of this project
- [x] 👀 I have performed a self-review of my own code
- [ ] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [x] ⬆️ I increased versions for any altered app according to semantic versioning

**➕ App addition**

If this PR is an app addition please make sure you have done the following.

- [ ] 🪞 I have opened a PR on [truecharts/containers](https://github.com/truecharts/containers) adding the container to TrueCharts mirror repo.
- [ ] 🖼️ I have added an icon in the Chart's root directory called `icon.png`

---

_Please don't blindly check all the boxes. Read them and only check those that apply.
Those checkboxes are there for the reviewer to see what is this all about and
the status of this PR with a quick glance._

### CODE REVIEW

**Comment by stavros-k** [comment]
It now seems to generate it correctly.

But few things need to be checked. Might be an issue because I don't understand something

- There are references in `nameOverride` under both `rbac` and `serviceAccount` 
  but only `name` key exists in `values.yaml` as example.
  Q: Are those 2 different things?

- Should it be better to move SA name generation into a lib instead of duplicating basically what I have added in this PR,    
  each time we need to use SA name somewhere?

-  In the names.tpl there are references of `serviceAccount.create`, afaik with the refactor we changed it to 
   `serviceAccount.main.enable` (?) Not really sure about that.

**Comment by Ornias1993** [comment]
> * There are references in `nameOverride` under both `rbac` and `serviceAccount`
>   but only `name` key exists in `values.yaml` as example under `serviceAccount` and neither under `rbac`
>   Q: Are those 2 different things?

No, see the spawner.

> * Should it be better to move SA name generation into a lib instead of duplicating basically what I have added in this PR,
>   each time we need to use SA name somewhere?

It was already in a lib and all this PR does is remove that lib-reference to replace it with your thing.
Rather you update the lib

> * In the names.tpl there are references of `serviceAccount.create`, afaik with the refactor we changed it to
>   `serviceAccount.main.enable` (?) Not really sure about that.

That was indeed changed, however: i'm not sure that names.tpl actually is updated at all for our 'allow multiple SA' update

File: charts/common/templates/lib/chart/_names.tpl [hunk 1]
```diff
@@ -37,8 +37,12 @@ If release name contains chart name it will be used as a full name.
...
 {{/* Create the name of the ServiceAccount to use */}}
 {{- define "tc.common.names.serviceAccountName" -}}
-  {{- if .Values.serviceAccount.create -}}
-    {{- default (include "tc.common.names.fullname" .) .Values.serviceAccount.name -}}
+  {{- if .Values.serviceAccount.enabled -}}
+  {{- $saName := (include "tc.common.names.fullname" .) -}}
+    {{- if and (hasKey .Values.serviceAccount "nameOverride") .Values.serviceAccount.nameOverride -}}
+      {{- $saName = printf "%v-%v" $saName .Values.serviceAccount.nameOverride -}}
+    {{- end -}}
+    {{- default $saName .Values.serviceAccount.name -}}
```
Comment (stavros-k) [inline charts/common/templates/lib/chart/_names.tpl:45]: This needs fixing...
I think I'm passing wrong values from class.

As it always ends up in the "default" case

Comment (Ornias1993) [reply charts/common/templates/lib/chart/_names.tpl:45]: Okey, some info:
When multiple things are spawned, they get read and put into a seperate object by the spawner.
This object is used from this point onwards by the class.

You **CANNOT** reach into `.Values` for ANYTHING related to the object.

Also: `.Values.serviceAccount "nameOverride"` does not exist at all.
Even if it would work it would be `.Values.somename.serviceAccount "nameOverride"`.
But that cannot be used, you CANNOT reference anything using `.Values` in this class. Everything needs to be inhereted by the include. If you cannot do that, you will need to get rid of the lib, basically. (which i think is best, because you honestly cannot use this include anywhere the way it needs to be writen to even work on SA.

Comment (Ornias1993) [reply charts/common/templates/lib/chart/_names.tpl:45]: So TLDR:
Nuke the idea and do not use a lib for this, you cannot get it to work without making a mess out of everything.

Comment (Ornias1993) [reply charts/common/templates/lib/chart/_names.tpl:45]: So: Get rid of this lib and define the SA names on 2 classes (SA and CRB) manually

File: charts/common/templates/lib/controller/_pod.tpl [hunk 2]
```diff
@@ -6,7 +6,7 @@ The pod definition included in the controller.
 imagePullSecrets:
     {{ tpl ( toYaml . ) $ | nindent 2 }}
   {{- end }}
-serviceAccountName: {{ include "tc.common.names.serviceAccountName" . }}
+serviceAccountName: "default"
```
Comment (stavros-k) [inline charts/common/templates/lib/controller/_pod.tpl:9]: @Ornias1993  With the lib removal, this also needs updating. 
I changed it to `default`, but I have no idea what this should be set to.

**Comment by stavros-k** [comment]
Given this:

```yaml
serviceAccount:
  main:
    enabled: true
  main2:
    enabled: true

rbac:
  main:
    enabled: true
  main2:
    enabled: true
```


Result for SA's is:
````yaml
# Source: portainer-ce/templates/common.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
✅
  name: release-name-portainer-ce
✅
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
---
# Source: portainer-ce/templates/common.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
✅
  name: release-name-portainer-ce-main2
✅ 
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
````

Result for ClusterRoleBinding's is:
````yaml
# Source: portainer-ce/templates/common.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: release-name-portainer-ce
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
  annotations:
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
✅
  name: release-name-portainer-ce
✅
subjects:
  - kind: ServiceAccount
✅  
    name: release-name-portainer-ce
✅
    namespace: default
---
# Source: portainer-ce/templates/common.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: release-name-portainer-ce-main2
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
  annotations:
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
✅
  name: release-name-portainer-ce-main2
✅  
subjects:
  - kind: ServiceAccount
❌  
    name: release-name-portainer-ce
❌
    namespace: default
````

**Comment by stavros-k** [comment]
Now it SA name under subject is correct (when multiple defined)

Assuming rbac and sa have the same name.

**Comment by stavros-k** [comment]
Given this:

````yaml
serviceAccount:
  main:
    enabled: true
  main2:
    enabled: true

rbac:
  main:
    enabled: true
->
  main3:
    enabled: true
````


Result for SA's is:
````yaml
# Source: portainer-ce/templates/common.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
✅
  name: release-name-portainer-ce
✅
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
---
# Source: portainer-ce/templates/common.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
✅
  name: release-name-portainer-ce-main2
✅ 
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
````

Result for ClusterRoleBinding's is:
````yaml
# Source: portainer-ce/templates/common.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: release-name-portainer-ce
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
  annotations:
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
✅
  name: release-name-portainer-ce
✅
subjects:
  - kind: ServiceAccount
✅  
    name: release-name-portainer-ce
✅
    namespace: default
---
# Source: portainer-ce/templates/common.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: release-name-portainer-ce-main2
  labels:
    helm.sh/chart: portainer-ce-0.0.1
    app.kubernetes.io/name: portainer-ce
    app.kubernetes.io/instance: release-name
    helm-revision: "1"
    app.kubernetes.io/version: "2.15.1"
    app.kubernetes.io/managed-by: Helm
  annotations:
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
✅
  name: release-name-portainer-ce-main3
✅  
subjects:
  - kind: ServiceAccount
❓ 
    name: default
❓  
    namespace: default
````

File: charts/common/templates/lib/controller/_pod.tpl (hunk 2 above)
Comment (Ornias1993) [reply charts/common/templates/lib/controller/_pod.tpl:9]: To the name of the main SA

**Comment by Ornias1993** [comment]
TLDR:
Default == main SA

additional SA’s are mostly usefull for additional pods.

**Review by Ornias1993 — APPROVED** [review]
Some important notes:
Normally every pod defaults to the "default" service account in any namespace.
This change means we will say "if we create our own, use that instead".

Security wise, this should prove to be considerably more relyable, as it prevents conflicts when multiple pods/charts are installed into the same namespace.
